File: backend/internal/engine/bitboard.go
package engine

import (
	"caro-ai-pvp/internal/domain"
	"math/bits"
)

type BitBoard [4]uint64

func bitIndex(x, y int) (int, uint) {
	idx := y*domain.BoardSize + x
	return idx >> 6, uint(idx & 63)
}

func (b *BitBoard) Set(x, y int) {
	i, off := bitIndex(x, y)
	b[i] |= 1 << off
}

func (b *BitBoard) Clear(x, y int) {
	i, off := bitIndex(x, y)
	b[i] &^= 1 << off
}

func (b BitBoard) Get(x, y int) bool {
	i, off := bitIndex(x, y)
	return b[i]&(1<<off) != 0
}

func (b BitBoard) Or(other BitBoard) BitBoard {
	return BitBoard{b[0] | other[0], b[1] | other[1], b[2] | other[2], b[3] | other[3]}
}

func (b BitBoard) And(other BitBoard) BitBoard {
	return BitBoard{b[0] & other[0], b[1] & other[1], b[2] & other[2], b[3] & other[3]}
}

func (b BitBoard) Xor(other BitBoard) BitBoard {
	return BitBoard{b[0] ^ other[0], b[1] ^ other[1], b[2] ^ other[2], b[3] ^ other[3]}
}

func (b BitBoard) Not() BitBoard {
	return BitBoard{^b[0], ^b[1], ^b[2], ^b[3]}
}

func (b BitBoard) IsZero() bool {
	return b[0] == 0 && b[1] == 0 && b[2] == 0 && b[3] == 0
}

func (b BitBoard) Count() int {
	return bits.OnesCount64(b[0]) + bits.OnesCount64(b[1]) +
		bits.OnesCount64(b[2]) + bits.OnesCount64(b[3])
}

func (b BitBoard) Dilate() BitBoard {
	const W = domain.BoardSize
	var result BitBoard

	// Iterate over all 256 bits, for each set bit set all 8 neighbors
	for y := range W {
		for x := range W {
			if !b.Get(x, y) {
				continue
			}
			for dy := -1; dy <= 1; dy++ {
				for dx := -1; dx <= 1; dx++ {
					nx, ny := x+dx, y+dy
					if nx >= 0 && nx < W && ny >= 0 && ny < W {
						result.Set(nx, ny)
					}
				}
			}
		}
	}
	return result
}

func BitBoardsFromDomain(b domain.Board) (red, blue BitBoard) {
	rBits := b.BitBoardBits(domain.PlayerRed)
	bBits := b.BitBoardBits(domain.PlayerBlue)
	copy(red[:], rBits[:])
	copy(blue[:], bBits[:])
	return
}

File: backend/src/Caro.Engine/BitBoard.cs
using Caro.Domain;

namespace Caro.Engine;

/// <summary>
/// Four-word bitboard over the 16x16 board. Value-type semantics like the
/// Go original: Set/Clear mutate the instance they are called on.
/// </summary>
public struct BitBoard
{
    private ulong _w0;
    private ulong _w1;
    private ulong _w2;
    private ulong _w3;

    private readonly ulong Word(int i) => i switch
    {
        0 => _w0,
        1 => _w1,
        2 => _w2,
        _ => _w3,
    };

    private void SetWord(int i, ulong v)
    {
        switch (i)
        {
            case 0: _w0 = v; break;
            case 1: _w1 = v; break;
            case 2: _w2 = v; break;
            default: _w3 = v; break;
        }
    }

    internal static (int Word, int Offset) BitIndex(int x, int y)
    {
        int idx = y * Constants.Board.Size + x;
        return (idx >> 6, idx & 63);
    }

    public void Set(int x, int y)
    {
        (int i, int off) = BitIndex(x, y);
        SetWord(i, Word(i) | (1UL << off));
    }

    public void Clear(int x, int y)
    {
        (int i, int off) = BitIndex(x, y);
        SetWord(i, Word(i) & ~(1UL << off));
    }

    public readonly bool Get(int x, int y)
    {
        (int i, int off) = BitIndex(x, y);
        return (Word(i) & (1UL << off)) != 0;
    }

    public readonly BitBoard Or(in BitBoard other) => new()
    {
        _w0 = _w0 | other._w0,
        _w1 = _w1 | other._w1,
        _w2 = _w2 | other._w2,
        _w3 = _w3 | other._w3,
    };

    public readonly BitBoard And(in BitBoard other) => new()
    {
        _w0 = _w0 & other._w0,
        _w1 = _w1 & other._w1,
        _w2 = _w2 & other._w2,
        _w3 = _w3 & other._w3,
    };

    public readonly BitBoard Xor(in BitBoard other) => new()
    {
        _w0 = _w0 ^ other._w0,
        _w1 = _w1 ^ other._w1,
        _w2 = _w2 ^ other._w2,
        _w3 = _w3 ^ other._w3,
    };

    public readonly BitBoard Not() => new()
    {
        _w0 = ~_w0,
        _w1 = ~_w1,
        _w2 = ~_w2,
        _w3 = ~_w3,
    };

    public readonly bool IsZero() => _w0 == 0 && _w1 == 0 && _w2 == 0 && _w3 == 0;

    public readonly int Count() =>
        (int)(ulong.PopCount(_w0) + ulong.PopCount(_w1) + ulong.PopCount(_w2) + ulong.PopCount(_w3));

    public readonly BitBoard Dilate()
    {
        const int W = Constants.Board.Size;
        BitBoard result = default;

        // Iterate over all 256 bits, for each set bit set all 8 neighbors
        for (int y = 0; y < W; y++)
        {
            for (int x = 0; x < W; x++)
            {
                if (!Get(x, y))
                {
                    continue;
                }
                for (int dy = -1; dy <= 1; dy++)
                {
                    for (int dx = -1; dx <= 1; dx++)
                    {
                        int nx = x + dx;
                        int ny = y + dy;
                        if (nx >= 0 && nx < W && ny >= 0 && ny < W)
                        {
                            result.Set(nx, ny);
                        }
                    }
                }
            }
        }
        return result;
    }

    public static (BitBoard Red, BitBoard Blue) BitBoardsFromDomain(Board b)
    {
        ulong[] rBits = b.BitBoardBits(Player.Red);
        ulong[] bBits = b.BitBoardBits(Player.Blue);
        return (
            new BitBoard { _w0 = rBits[0], _w1 = rBits[1], _w2 = rBits[2], _w3 = rBits[3] },
            new BitBoard { _w0 = bBits[0], _w1 = bBits[1], _w2 = bBits[2], _w3 = bBits[3] });
    }
}
